include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -260,4 +262,146 @@ def YieldOp : VerifOp<"yield", [
+      return RegionKind::Graph;
+    }
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Verification Ops
+//===----------------------------------------------------------------------===//
+
+def SymbolicInputOp : VerifOp<"symbolic_input", [HasParent<"verif::FormalOp">]>{
+  let summary = "declare a symbolic value for formal verification";
+  let description = [{
+    This operation declares a symbolic value that can then be used to reason 
+    about `hw.instance` inputs in a symbolic manner in future assertions and 
+    assumptions. The result type must be explicitly marked.  
+  }];
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = "attr-dict `:` type($result)";
+}
+
+def ConcreteValOp : VerifOp<"concrete_input", [
+  AllTypesMatch<["init", "update"]>, HasParent<"verif::FormalOp">
+]> {
+  let summary = "declare a concrete value for formal verification";
+  let description = [{
+    This operation declares a concrete value that can then be used in the 
+    reasoning surrounding symbolic values, allowing for a form of concolic 
+    reasoning to take place in a `verif.formal` block. Concrete values are 
+    defined by an initial value and an update function.
+  }];
+
+  let arguments = (ins AnyType:$init, AnyType:$update);
+  let results = (outs AnyType:$result);
+
+  // Manual definition of a functional type is wanted
+  let assemblyFormat = [{
+    $init `,` $update attr-dict `:` 
+    `(` type($init) `,` type($update) `)` `->` type($result)
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Contract Ops
+//===----------------------------------------------------------------------===//
+
+def ContractOp : VerifOp<"contract", [Pure]>{
+  let summary = "declare formal contract";
+  let description = [{
+    This operation declares a sformal contract that is used to create precondition 
+    and postconditions on a parent module. These are used as an abstraction to 
+    better modularize formal verification such that each module containing a contract 
+    is checked exactly once.
+  }];
+  let regions = (region SizedRegion<1>:$contract_region);
+
+  let assemblyFormat = [{
+    attr-dict-with-keyword $contract_region
+  }];
+}
+
+def RequireOp : VerifOp<"require", [HasParent<"verif::ContractOp">]>{
+  let summary = "define a precondition for the given contract";
+  let description = [{
+    This operation defines a precondition for the parent contract. 
+    Preconditions are assumed during the verification of a module and 
+    asserted during the verification of an instance of a module.
+  }];
+  let arguments = (ins LTLAnyPropertyType:$property,
+                      ClockEdgeAttr:$edge, I1:$clock,
+                      Optional<I1>:$enable,
+                      OptionalAttr<StrAttr>:$label);
+
+  let assemblyFormat = [{
+    $property (`if` $enable^)? `,` $edge $clock 
```

Comment:
Similarly, is the clock attribute to allow the conditions to only apply on a given clock edge? If so, do we know how this lowers to our backends (e.g. can btor2 encode checking a property on a clock negedge?)

Reply:
the idea with using enable signals and clocks is that these should be able to support SVA properties, which are inherently invalid before the a given state has been reached, and this needs to be specified somewhere, e.g. `a |=> b` is invalid until the system has been reset at least once since a and b both have garbage values at that point, and that condition needs to be defined somewhere. 

Basically I want to give the same modeling power as writing assertions and assumptions by hand